packages/entity-server/src/routes/hierarchy/EntityRelationsRoute.ts

```diff
@@ -0,0 +1,150 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { Route } from '@tupaia/server-boilerplate';
+import { reduceToDictionary, reduceToArrayDictionary } from '@tupaia/utils';
+import { EntityType } from '../../models';
+import {
+  HierarchyRequest,
+  HierarchyRequestParams,
+  HierarchyRequestBody,
+  HierarchyRequestQuery,
+  EntityResponseObject,
+} from './types';
+
+export type RelationsRequest = HierarchyRequest<
+  HierarchyRequestParams,
+  | (EntityResponseObject & { descendants: EntityResponseObject[] })[] // groupBy: ancestor, flattenToCode: false
+  | (EntityResponseObject & { ancestor: EntityResponseObject })[] // groupBy: descendant, flattenToCode: false
+  | Record<string, string[]> // groupBy: ancestor, flattenToCode: true
+  | Record<string, string>, // groupBy: descendant, flattenToCode: true
+  HierarchyRequestBody,
+  HierarchyRequestQuery & {
+    ancestorType?: string;
+    descendantType?: string;
+    groupBy?: 'ancestor' | 'descendant';
+    flattenToCode?: 'true' | 'false';
+  }
+>;
+
+export class EntityRelationsRoute extends Route<RelationsRequest> {
+  async buildDescendantAncestorPairs(descendants: EntityType[], ancestorType: string) {
+    const { hierarchyId } = this.req.ctx;
+    const descendantAncestorMapping = await this.req.models.entity.fetchAncestorDetailsByDescendantCode(
+      descendants.map(descendant => descendant.code),
+      hierarchyId,
+      ancestorType,
+    );
+
+    return Object.entries(descendantAncestorMapping).map(([descendant, { code: ancestor }]) => ({
+      descendant,
+      ancestor,
+    }));
+  }
+
+  async buildFormattedEntitiesByCode(ancestors: EntityType[], descendants: EntityType[]) {
+    const allEntities = [...ancestors, ...descendants];
```

Comment:
Combining ancestors and descendants should be the caller's responsibility:

```suggestion
  async buildFormattedEntitiesByCode(entities: EntityType[]) {
```